i18n loader script: load script in the footer and defer it.

jeherve opened the pull request:
## Proposed changes:

Let's load that script in the footer of the page to improve performance.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

Related conversation: p55Cj4-3qL-p2

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

* Start by switching your site's language to French in Settings > General.
* Go to Dashboard > Updates and ensure that there are no pending translation updates.
* Go to Jetpack > Dashboard > My Plans
* Purchase a Jetpack Search Plan
* Back on the site, go to Jetpack > Settings > Performance
* Enable Search and Instant Search
* Go to Appearance > Editor
* Add a Search block somewhere on your homepage.
* Load your site's frontend.
* Click on the search field and launch a search.
* The Instant Search modal should pop up.
* The different UI elements in the modal should be in French.
* Now check the site's source code, and search for `wp-jp-i18n-loader`
* The script should be loaded close to the footer of the page.

To test the changes to `react-jsx-runtime`, you'll need to run through the testing instructions in #38428.

jeherve committed:
i18n loader script: load script in the footer and defer it.

Related conversation: p55Cj4-3qL-p2

jeherve committed:
Add 'add_data' mock

anomiex reviewed (dismissed):
This works, although it doesn't look like there's much point to the "defer" part since [the existence of an "after" script will always disable it](https://github.com/WordPress/wordpress-develop/blob/13d3c5ee0821ab6b8a8c91ff88f79a829e85de7f/src/wp-includes/class-wp-scripts.php#L957-L960).

anomiex reviewed (commented):
Also, it might be worth doing similarly for the 'react-jsx-runtime' registered just below this.

anomiex commented on projects/packages/assets/tests/php/test-assets.php:
```diff
@@ -729,7 +729,7 @@ function ( $v ) {
 		// @phan-suppress-next-line PhanDeprecatedFunction -- Keep using setMethods until we drop PHP 7.0 support.
 		$mock = $this->getMockBuilder( \stdClass::class )
-			->setMethods( array( 'add', 'add_inline_script' ) )
+			->setMethods( array( 'add', 'add_inline_script', 'add_data' ) )
```
Should we add some `$mock->expects( ... )->method( 'add_data' )` to check that the proper calls were made for that too?

jeherve replied:
Sure thing, good idea. Done in 1503ba497019f68c2b381b0d2fb2341dfb2326bf

jeherve committed:
Remove defer strategy

It won't be needed:
https://github.com/Automattic/jetpack/pull/38929#pullrequestreview-2242806699

jeherve committed:
Also add react-jsx-runtime to be loaded in footer

See https://github.com/Automattic/jetpack/pull/38929#pullrequestreview-2242815676

Co-authored-by: Brad Jorsch <brad.jorsch@automattic.com>

jeherve committed:
Add test to ensure add_data is being called

See https://github.com/Automattic/jetpack/pull/38929#discussion_r1719904501

jeherve commented:
> It doesn't look like there's much point to the "defer" part since [the existence of an "after" script will always disable it](https://github.com/WordPress/wordpress-develop/blob/13d3c5ee0821ab6b8a8c91ff88f79a829e85de7f/src/wp-includes/class-wp-scripts.php#L957-L960).

Good point. Removed in ec10133d39c4fe1aecc7a76db5def190d89de0e5

> Also, it might be worth doing similarly for the 'react-jsx-runtime' registered just below this.

👍 Done in 82fb0aac0238fa3e52a13be2205318242a2fa9ad

anomiex commented on projects/packages/assets/changelog/update-i18n-loader-footer:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: changed
+
+i18n loader script: load script in the footer and defer it.
```
Update this?

jeherve replied:
Good point. 2c1de55532b8a1696de827c530c98b48d65b8549

jeherve committed:
Update changelog
